crates/sui-core/src/authority/tally_rule.rs

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use rocksdb::{DBWithThreadMode, MultiThreaded, Options};
+use std::collections::BTreeMap;
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+use sui_types::base_types::{AuthorityName, ExecutionDigests};
+use sui_types::error::SuiResult;
+use typed_store::rocks::DBMap;
+use typed_store::{reopen, Map};
+
+const TX_EFFECTS_GOSSIP_ORDER: &str = "tx_effects_gossip_order";
+const TALLY_SCORES: &str = "tally_scores";
+
+pub type GossipSequenceNumber = u64;
+
+pub struct TallyRecord {
+    next_seq: AtomicU64,
+    /// Stores the names of validators that sent us execution digests through gossip.
+    /// The key of this table is a tuple of the execution digest and the name of the validator.
+    /// The value is a sequence number obtained from an atomic counter that increase each time
+    /// we receive an item from gossip. It can be used to determine the order we received from
+    /// each validator for each transaction.
+    /// At the end of each checkpoint, we go through each digest executed in the checkpoint,
+    /// update the score of validators based on the order we received from them. We also remove
+    /// all the digests processed.
+    /// In case there are leaks (i.e. digests added back after processing), at the end of epoch,
+    /// we clear this table.
+    tx_effects_gossip_order: DBMap<(ExecutionDigests, AuthorityName), GossipSequenceNumber>,
```

Comment:
Since validators follow a randomly chosen subset of other validators for gossip, this is highly sensitive to that randomness. A validator that just happened not to be chosen as a gossip peer by anyone else will be punished through no fault of their own.

Possibly, over the course of an epoch, this will even out, but that's not obvious to me.

Reply:
Indeed we are relying on probability here, which is not guaranteed to be fair.
One thing we could do is to take a note of the validators that are selected as peers each time, and use that to normalize the scores